File: client/app/src/main/java/com/unina/natour/domain/LoginUseCase.java
package com.unina.natour.domain;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.amazonaws.mobileconnectors.cognitoauth.Auth;
import com.amplifyframework.auth.AuthProvider;
import com.amplifyframework.core.Amplify;
import com.unina.natour.ui.RootActivity;

import javax.inject.Inject;

public class LoginUseCase {

    @Inject
    LoginUseCase() {
    }

    public MutableLiveData<Boolean> login(String username, String password) {
        MutableLiveData<Boolean> loginState = new MutableLiveData<>();

        Amplify.Auth.signIn(
                username,
                password,
                result -> {
                    Log.i("AuthQuickstart",
                            result.isSignInComplete() ? "Sign in succeeded" : "Sign in not complete");

                    loginState.postValue(result.isSignInComplete());
                },
                error -> {
                    Log.e("AuthQuickstart", error.toString());

                    loginState.postValue(false);
                }
        );

        return loginState;
    }
}

File: client/app/src/main/java/com/unina/natourkt/domain/usecase/LoginUseCase.kt
package com.unina.natourkt.domain.usecase

import android.content.Context
import android.util.Log
import androidx.datastore.core.CorruptionException
import com.amplifyframework.auth.AuthException
import com.unina.natourkt.R
import com.unina.natourkt.common.DataState
import com.unina.natourkt.common.ErrorHandler
import com.unina.natourkt.data.remote.dto.toUser
import com.unina.natourkt.domain.repository.AuthRepository
import com.unina.natourkt.domain.repository.DataStoreRepository
import com.unina.natourkt.domain.repository.UserRepository
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import retrofit2.HttpException
import java.io.IOException
import java.lang.Exception
import javax.inject.Inject

/**
 * This UseCase make use of
 * - [AuthRepository] to login the user
 * - [DataStoreRepository] to persist the user on DataStore Preferences
 * - [UserRepository] to retrieve the user through REST Service
 */
class LoginUseCase @Inject constructor(
    private val authRepository: AuthRepository,
    private val dataStoreRepository: DataStoreRepository,
    private val userRepository: UserRepository,
    private val errorHandler: ErrorHandler,
) {

    operator fun invoke(username: String, password: String): Flow<DataState<Boolean>> = flow {
        try {
            emit(DataState.Loading())

            val isSignInComplete = authRepository.login(username, password)

            if (isSignInComplete) {
                val user = userRepository.getUserByCognitoId(authRepository.fetchUserSub()).toUser()
                dataStoreRepository.saveUserToDataStore(user);

                emit(DataState.Success(isSignInComplete))
            } else {
                emit(DataState.Error(DataState.CustomMessages.SomethingWentWrong("Unknown Error")))
            }
        } catch (e: Exception) {
            emit(DataState.Error(errorHandler.handleException<Throwable>(e)))
        }
    }
}